### ISSUE TITLE
Android App Crashes after some time

### ISSUE THREAD

**Issue by drexhacker** [781flyingdutchman/background_downloader#268]
**Describe the bug**
Hello thanks for the great work, the downloader actually works well but some times it makes the app to crash randomly.

I first noticed it when I only had a few downloads (3). And Actually today I was supprised it downloaded upto 1000 + files but still it crashed somewhere.

**To Reproduce**
I am not sure what to put here but I am just using `downloadBatch` and testing on my Android 8 phone and Android 14 emulator.
And on desktop Everything just works fine.

**Expected behavior**
I personally I think if the worker crashes, it shouldn't crash the app also.

**Logs**
If possible, include logs that capture the issue:
```
I/WM-WorkerWrapper(19167): Worker result SUCCESS for Work [ id=45a8e488-4d0f-44d5-914b-846ba2c4d6ad, tags={ com.bbflight.background_downloader.DownloadTaskWorker, BackgroundDownloader, taskId=2941648995, group=default } ]
I/WM-WorkerWrapper(19167): Worker result SUCCESS for Work [ id=776a3b8a-f600-451d-82da-16cc33e36aa8, tags={ com.bbflight.background_downloader.DownloadTaskWorker, BackgroundDownloader, taskId=2861137335, group=default } ]
E/AndroidRuntime(19167): FATAL EXCEPTION: DefaultDispatcher-worker-2
E/AndroidRuntime(19167): Process: com.hsecampus.app, PID: 19167
E/AndroidRuntime(19167): java.lang.IllegalStateException: Calls to setForegroundAsync() must complete before a ListenableWorker signals completion of work by returning an instance of Result.
E/AndroidRuntime(19167):        at androidx.work.impl.utils.WorkForegroundUpdater$1.run(WorkForegroundUpdater.java:95)
E/AndroidRuntime(19167):        at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
E/AndroidRuntime(19167):        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/AndroidRuntime(19167):        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/AndroidRuntime(19167):        at java.lang.Thread.run(Thread.java:920)
E/AndroidRuntime(19167):        Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@a4c5096, Dispatchers.Default]
D/EGL_emulation(19167): app_time_stats: avg=244016.88ms min=244016.88ms max=244016.88ms count=1
I/Process (19167): Sending signal. PID: 19167 SIG: 9
```

**Code**
If possible, paste in the code snippet where you interact with the package:
```

  /// Downloads File to the Application Support Directory
  ///
  /// It skips Files that already exist on the system
  void downloadFiles(List<String> urls, {bool checked = false}) async {
    // Initialize empty download task list
    List<DownloadTask> tasks = [];
    // Check if Some of the files already Exist and skip them
    for (int i = 0; i < urls.length; i++) {
      final fileUri = Uri.parse(urls[i]);
      // if `checked` then don't check file exists
      if (checked) {
        tasks.add(
          DownloadTask(
            url: fileUri.toString(),
            filename: p.basename(fileUri.path),
            directory: p.dirname(fileUri.path),
            baseDirectory: BaseDirectory.applicationSupport,
            allowPause: true,
          ),
        );
      } else {
        if (!File("${_baseDir!}/${fileUri.path}").existsSync()) {
          tasks.add(
            DownloadTask(
              url: fileUri.toString(),
              filename: p.basename(fileUri.path),
              directory: p.dirname(fileUri.path),
              baseDirectory: BaseDirectory.applicationSupport,
              allowPause: true,
            ),
          );
        } else {
          // debugPrint("File: ${fileUri.path} already exists.");
        }
      }
    }
    // Start Downloading the Files
    if (tasks.isNotEmpty) {
      _downloader.downloadBatch(
        tasks,
        batchProgressCallback: (succeeded, failed) {
          debugPrint('$succeeded files succeeded, $failed have failed');
        },
        taskStatusCallback: (update) async {
        },
      );
    }
  }
```

**Additional context**
Nothing Really.

**Comment by 781flyingdutchman** [comment]
Thanks for sharing. Can you confirm:
* Your targetSdk is 34
* You have used Config.runInForeground

The issue seems related to the fact that the notification updates are queued, and in some cases the update only gets done when the task has already finished. I'll need to look into it further and may ask for your help testing it if that's ok.

**Comment by 781flyingdutchman** [comment]
I think I may have the fix - can you confirm this works for you? Change your pubspec.yam entry to:
```
background_downloader:
    git:
      url: https://github.com/781flyingdutchman/background_downloader.git
      ref: foreground
```
to use the modified (unpublished) version and let me know if the problem has disappeared. Thanks!

**Comment by drexhacker** [comment]
> I think I may have the fix - can you confirm this works for you? Change your pubspec.yam entry to:
> 
> ```
> background_downloader:
>     git:
>       url: https://github.com/781flyingdutchman/background_downloader.git
>       ref: foreground
> ```
> 
> to use the modified (unpublished) version and let me know if the problem has disappeared. Thanks!

Hello @781flyingdutchman Sorry for my late response let me apply this immediately and get back to you.

> Thanks for sharing. Can you confirm:
> 
> * Your targetSdk is 34
> * You have used Config.runInForeground
> 
> The issue seems related to the fact that the notification updates are queued, and in some cases the update only gets done when the task has already finished. I'll need to look into it further and may ask for your help testing it if that's ok.

This is also true.

**Comment by drexhacker** [comment]
Hello after applying your changes I have the following error.
```
e: file:///home/admin/.pub-cache/git/background_downloader-dbcac42a300f3f55360de9821a14a51a642ab2f9/android/src/main/kotlin/com/bbflight/background_downloader/Notifications.kt:308:56 Unresolved reference: UpdateNotificationWorker
e: file:///home/admin/.pub-cache/git/background_downloader-dbcac42a300f3f55360de9821a14a51a642ab2f9/android/src/main/kotlin/com/bbflight/background_downloader/Notifications.kt:313:32 Unresolved reference: UpdateNotificationWorker

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':background_downloader:compileDebugKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details
```

**Comment by 781flyingdutchman** [comment]
I think this may be a cache issue: run `flutter clean` and try again?

**Comment by drexhacker** [comment]
@781flyingdutchman I haven't seen this Error again. Hope it is resolved now. Thank You So much.

**Comment by drexhacker** [comment]
> I think this may be a cache issue: run `flutter clean` and try again?

Yeah it was cache.

**Comment by 781flyingdutchman** [comment]
Implemented in V8.4.1
